src/devinterp/slt/mala.py

```diff
@@ -0,0 +1,130 @@
+    prev_loss: Loss of the previous point in parameter space.
+    current_grad: Gradient of the current point in parameter space.
+    current_point: The current point in parameter space.
+    current_loss: Loss of the current point in parameter space.
+    learning_rate (float): Learning rate of the model.
+
+    Returns:
+    float: Acceptance probability for the proposed transition.
+    """
+    if torch.isnan(current_loss):
+        return np.nan
+    # Compute the log of the proposal probabilities (using the Gaussian proposal distribution)
+    log_q_current_to_prev = -torch.sum(
+        (prev_point - current_point - (learning_rate * 0.5 * -current_grad)) ** 2
+    ) / (2 * learning_rate)
+    log_q_prev_to_current = -torch.sum(
+        (current_point - prev_point - (learning_rate * 0.5 * -prev_grad)) ** 2
+    ) / (2 * learning_rate)
+    acceptance_log_prob = (
+        log_q_current_to_prev - log_q_prev_to_current + prev_loss - current_loss
+    )
+    return min(1.0, torch.exp(acceptance_log_prob))
+
+
+class MalaAcceptanceRate(SamplerCallback):
+    """
+    Callback for computing the norm of the gradients of the optimizer / sampler.
+
+    Attributes:
+        num_draws (int): Number of samples to draw. (should be identical to param passed to sample())
+        num_chains (int): Number of chains to run. (should be identical to param passed to sample())
+        num_samples (int): size of dataset passed to sample()
+        learning_rate (int): Learning rate of the model.
+        device (Union[torch.device, str]): Device to perform computations on, e.g., 'cpu' or 'cuda'.
+    """
+
+    def __init__(
+        self,
+        num_chains: int,
+        num_draws: int,
+        num_samples: int,
+        learning_rate: float,
+        device: Union[torch.device, str] = "cpu",
+    ):
+        self.num_chains = num_chains
+        self.num_draws = num_draws
+        self.learning_rate = learning_rate
+        self.num_samples = num_samples
+        self.mala_acceptance_rate = torch.zeros(
+            (num_chains, num_draws - 1), dtype=torch.float32
+        ).to(device)
+        self.device = device
+        self.current_params = []
+        self.current_grads = []
+        self.prev_params = []
+        self.prev_grads = []
+        self.prev_mala_loss = 0.0
+
+    def __call__(self, chain: int, draw: int, model: nn.Module, loss: float, optimizer):
+        self.update(chain, draw, model, loss, optimizer)
+
+    def update(self, chain: int, draw: int, model: nn.Module, loss: float, optimizer):
+        # we need the grads & loss from the pass, but the current params are from after the step
+        # (so we update those only that after the calculation)
+        self.current_grads = optimizer.dws
+        # mala acceptance loss is different from pytorch supplied loss
+        mala_loss = (
+            loss * self.num_samples / np.log(self.num_samples)
+        ) + optimizer.elasticity_loss
```

Comment:
`optimizer.elasticity_loss` looks like it's stored as a torch.tensor of shape `(1,)`. I think the other terms in `mala_loss` here are floats, so I think this would convert mala_loss to a torch.tensor of shape `(1,)` as well? but it looks like we might want this to be a float (e.g. prev_mala_loss defaults to a float, not a tensor). lmk if i've misinterpreted this though